### DISCUSSION TITLE
Ketcher 2.2.0 (April 15, 2021)

### DISCUSSION THREAD

**Discussion by AndreiMazol** [epam/ketcher#481] [General]
**Summary**

We are glad to announce that a new stable version of Ketcher 2.2.0 is released. The main goal of current release was adding support of standalone mode. The list of changes related to this functionality can be found [here](https://github.com/epam/ketcher/milestone/3). 

**New features and improvements** 

* Get Molfile in v3000 format #21 

* Track if origin file was changed #246 

* Add save to MDL molfile v3000 support #190

* Transform global css selectors into css classes #278

* Split Ketcher into Core and Presentations modules #196

* Extend editor API to generate png files based on structure #240

* Fire custom event on structure change #54

* Add delete button for user templates #160

* Add ketcherWindow.ketcher.getSmilesExt() method #266

* Add possibility to hide buttons on control panel #244

* Add bonds information for ket format in DAT s-groups #280

* Add indicator to inform users about execution of 'heavy' operations #255

* Ketcher API: Add possibility to save RXN files #296

* Add ellipse simple object #85

* 'Font' field should be disabled untill list will be loaded #321

* Modal window displaying in Ketcher instance format #325

* Simple Objects selection and resizing mechanism #315

* Add ability to download image in several formats #373

* Allow select simple object by several points in selection area #367

**Bugfixes**

* If structure contains SRU-Group User can't click 'Calculated Values' button - nothing happens after clicking. #215

* Stereo labels appears on not stereo structure after 'Aromatize' action. #206

* User can create incorrect Multiple S-Group but can't save and warning message doesn't appears. #175

* Calculated Values displays some values from last calculation #279

* Remove process.env.PUBLIC_URL from ketcher-react #282

* Fonts drop-down list can't be opened when User opens settings for the second time #320

* User can't copy/cut and paste part of structure or structure with SRU-polymer S-Group. #177

* When User draw circle without height - Ketcher becomes disabled. #317

* Error after Undo/Redo actions when User has edited S-Group #311

* Unable to calculate values for selected structure #361

* S-Group brackets place in incorrect place on structure with double bonds and sprouted single bonds #269

* Calculated values for reaction: missing last bracket for chemical formula #364

* The structures with atoms with valences rendering incorrectly after saving #46

* Different structure between ketcher.setMolecule() and ketcher.ui.load() #386

* Templates in template library are shifted down when Ketcher have less than 855px width #338

* Display Settings: 'Show hydrogen labels' triggers error #391

* 'Calculated values' fields are editable #388

* Incorrect behavior of Ketcher after deleting of a part of R-Group member #419

* In Template Library scroll bar jumps up and down when it is centered #184

* Incorrect behavior of Ketcher after deleting of a part of S-Group member #433

* 'Any Atom' tool incorrect behavior when selected with hotkey or typed in from keyboard #368

* 'Custom Templates' - User templates are shown without the images in the right pane of the 'Template library' #431

<hr /><em>This discussion was created from the release <a href='https://github.com/epam/ketcher/releases/tag/v2.2.0'>Ketcher 2.2.0 (April 15, 2021)</a>.</em>

**Comment by robtree** [comment]
Congratulations on the latest release!

In Ketcher 2.2.0 the index.html file now loads static assets from the root `"/"` rather than starting in the current directory `"./"`. I've had to change these paths back to `"./"` in order to keep Ketcher in its own `/public/ketcher` folder in a Vue.js project, as I have with Ketcher 2.1.0. Still, the `/docs/help.md` file must be kept in the root as, unlike `api_path`, it doesn't appear possible to override `staticResourcesUrl` in the `<iframe>` `src` attribute.

Is the change from `"./"` to `"/"` intentional, and is there a better way to load Ketcher in an `<iframe>` when the Ketcher application isn't stored in the root directory?

**Reply by AndreiMazol** [reply]
Hi @robtree ,

ideally ketcher should be hosted as an independent application in case it uses in 'IFrame' mode so '.' symbol is not required. Ketcher example is based on CRA template so you can update 'PUBLIC_URL' and create your own production build (please look at [this article](https://dev.to/n1ru4l/configure-the-cra-public-url-post-build-with-node-js-and-express-4n8) for example or use official CRA documentation). Another option is not to use Ready-to-use application but create your own react app that will use ketcher-react npm package (it gives you more flexibility such as ability to hide buttons and etc.) but it requires more time to  implement.

**Reply by robtree** [reply]
Hi, thank you @AndreiMazol.

Ketcher is hosted independently of our Vue.js app (we are not building a React app). Placing it in the [public folder](https://cli.vuejs.org/guide/html-and-static-assets.html#the-public-folder) and loading `<iframe src="/ketcher/ketcher.html?api_path=/api/ketcher">` works well with 2.1.0.

![image](https://user-images.githubusercontent.com/3150192/115251361-6297af00-a122-11eb-806d-d8f4d31fd3e1.png)

The workaround is to put Ketcher 2.2.0 assets directly in the root `/public` folder rather than their own `/public/ketcher` folder but this is not as clean as 2.1.0.

**Reply by AndreiMazol** [reply]
I mean that our ready-to-run build is based on assumption  that  all resources will be placed in public folder (docs, templates and static folder). if you  want to use the structure that is provided on the picture above, you need to build you own Ketcher production build (as I said before).  Please follow the next steps to create it:

1. navigate to the root folder of ketcher repository
2.  run `yarn build`
3.  navigate to example folder
4.  run `yarn cross-env MODE=remote PUBLIC_URL=ketcher  react-app-rewired build`
5.  copy all resources into /{your-public-folder}/ketcher folder

In my case the structure is the following (ketcher index.html is placed to /ketcher folder too):
![image](https://user-images.githubusercontent.com/72735611/115253375-ff634800-a134-11eb-8c4e-885a5ffd11d3.png)

After that Ketcher app is available by the following link:
`http://localhost:5000/ketcher` (to test my scenario run `serve build` after step 5):
![image](https://user-images.githubusercontent.com/72735611/115253961-93cdaa80-a135-11eb-8f19-ceaf359df673.png)

**Reply by robtree** [reply]
Custom build looks good with `PUBLIC_URL=/ketcher` so thanks for the steps!

![image](https://user-images.githubusercontent.com/3150192/115258809-39c6e800-a129-11eb-83cc-e998f17c9d3c.png)

**Reply by AndreiMazol** [reply]
Dear @robtree ,

we decided to implement this solution by default so  I guess it saves your time when you will switch to the next Ketcher release. all changes  are implementing in scope of https://github.com/epam/ketcher/issues/508
